Description: Policy asks for gcc -02
Forwarded: not-needed
Author: Jérémy Lal <kapouer@melix.org>
Last-Update: 2011-02-16
--- a/SConstruct
+++ b/SConstruct
@@ -94,7 +94,7 @@ LIBRARY_FLAGS = {
       'CPPDEFINES':   ['ENABLE_DISASSEMBLER', 'DEBUG'],
     },
     'mode:release': {
-      'CCFLAGS':      ['-O3', '-fomit-frame-pointer', '-fdata-sections',
+      'CCFLAGS':      ['-O2', '-fomit-frame-pointer', '-fdata-sections',
                        '-ffunction-sections'],
     },
     'os:linux': {
